// @walrus/plugin-jest/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;
exports.defaultModes = void 0;

var _jest = require("jest");

var _defaultConfig = require("./defaultConfig.resolver");

var _customConfig = require("./customConfig.resolver");

var _jestConfigurationBuilder = require("./jestConfigurationBuilder");

const debug = require('debug')('walrus-plugin-jest');

function _default(api, conf) {
  api.registerCommand('test', {
    description: 'run unit tests with jest',
    usage: 'walrus test [options] <regexForTestFiles>',
    options: {
      '--watch': 'run tests in watch mode'
    },
    details: `All jest command line options are supported.\n` + `See https://facebook.github.io/jest/docs/en/cli.html for more details.`
  }, (args, rawArgv) => {
    const rootDir = api.resolve('.'); // 获取默认配置

    const defaultConfig = new _defaultConfig.DefaultConfigResolver(rootDir);
    const configuration = new _jestConfigurationBuilder.JestConfigurationBuilder(defaultConfig, new _customConfig.CustomConfigResolver()).buildConfiguration(process.cwd());
    rawArgv.push('--config', JSON.stringify(configuration)); // 未查找到测试文件正常退出

    rawArgv.push('--passWithNoTests');
    (0, _jest.run)(rawArgv).then(result => {
      debug(result);
    }).catch(e => {
      console.log(e);
    });
  });
}

const defaultModes = {
  test: 'test'
};
exports.defaultModes = defaultModes;
