// @jiumao/rc-table/src/rc-table.tsx
import React from 'react';
import classNames from 'classnames';
import { Table } from 'antd';
import {
  PaginationConfig,
  TableProps,
  SorterResult,
  TableCurrentDataSource
} from 'antd/es/table';
import './rc-table.less';

export interface ITableData<T> {
  list: T[];
  pagination?: PaginationConfig
}

interface IProps<T> extends TableProps<T> {
  onSelectRow?: (rows: T[]) => void;
  data: ITableData<T>;
  selectedRows?: T[];
  onChange?: (
    pagination: PaginationConfig,
    filters: Record<keyof T, string[]>,
    sorter: SorterResult<T>,
    extra?: TableCurrentDataSource<T>,
  ) => void;
}

const prefixCls = 'rc-table';

const RcTable: React.FC<IProps<any>> = (props) => {
  const {
    className,
    style,
    rowKey,
    data,
    onChange,
    ...restProps
  } = props;
  const { list = [], pagination } = data;

  const handleTableChange = (
    pagination,
    filters,
    sorter
  ) => {
    onChange && onChange(pagination, filters, sorter);
  };

  const paginationProps = {
    showSizeChanger: true,
    showQuickJumper: true,
    ...pagination,
  };

  return (
    <div
      className={classNames(className, {
        [`${prefixCls}`]: true
      })}
      style={style}
    >
      <Table
        rowKey={rowKey}
        dataSource={list}
        pagination={paginationProps}
        onChange={handleTableChange}
        {...restProps}
      />
    </div>
  )
};

RcTable.defaultProps = {
  rowKey: 'id',
  size: 'middle'
};

export default RcTable;
