// umi-build-dev/lib/plugins/commands/block/sdk/flagBabelPlugin/GUmiUIFlag.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _default = function _default(_ref) {
  var filename = _ref.filename,
      index = _ref.index,
      inline = _ref.inline,
      children = _ref.children;

  function clickHandler() {
    var el = document.getElementById('umi-ui-bubble');

    if (el && el.contentWindow) {
      el.contentWindow.postMessage(JSON.stringify({
        action: 'umi.ui.block.addBlock',
        payload: {
          filename: filename,
          index: index
        }
      }), '*');
    }
  }

  return _react.default.createElement("div", null, _react.default.createElement("div", {
    onClick: clickHandler,
    style: _objectSpread({}, inline ? {} : {
      margin: '10px 0',
      height: '60px',
      lineHeight: '60px'
    }, {
      background: 'rgba(24,144,255,0.15)',
      border: '1px dashed #1890ff',
      textAlign: 'center',
      color: '#329bff',
      fontSize: '14px',
      cursor: 'pointer',
      transition: 'all .3s',
      overflow: 'hidden',
      fontFamily: 'Chinese Quote,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,PingFang SC,Hiragino Sans GB,Microsoft YaHei,Helvetica Neue,Helvetica,Arial,sans-serif'
    }),
    className: "g_umiuiBlockAddEditMode"
  }, _react.default.createElement("svg", {
    width: "1em",
    height: "1em",
    fill: "currentColor",
    viewBox: "0 0 1024 1024",
    style: {
      display: 'inline-block',
      verticalAlign: 'middle',
      marginRight: '8px'
    }
  }, _react.default.createElement("path", {
    d: "M482 152h60c5.333 0 8 2.667 8 8v704c0 5.333-2.667 8-8 8h-60c-5.333 0-8-2.667-8-8V160c0-5.333 2.667-8 8-8zM176 474h672c5.333 0 8 2.667 8 8v60c0 5.333-2.667 8-8 8H176c-5.333 0-8-2.667-8-8v-60c0-5.333 2.667-8 8-8z",
    fill: "#329bff",
    fillRule: "nonzero"
  })), _react.default.createElement("span", {
    style: {
      display: 'inline-block',
      verticalAlign: 'middle'
    }
  }, "Add to here")), children);
};

exports.default = _default;
