// @walrus/shared-utils/lib/config-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = require("path");

var _fs = require("fs");

var _joycon = _interopRequireDefault(require("joycon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const requireFromString = require('require-from-string');

const configLoader = new _joycon.default({
  stopDir: (0, _path.dirname)(process.cwd())
});
configLoader.addLoader({
  test: /\.[jt]s$/,

  loadSync(id) {
    const content = require('@babel/core').transform((0, _fs.readFileSync)(id, 'utf8'), {
      babelrc: false,
      configFile: false,
      filename: id,
      presets: [[require('@babel/preset-env'), {
        targets: {
          node: 'current'
        }
      }], id.endsWith('.ts') && require('@babel/preset-typescript')].filter(Boolean)
    });

    const m = requireFromString(content && content.code ? content.code : '', id);
    return m.default || m;
  }

});
var _default = configLoader;
exports.default = _default;
